storage/innobase/fil/fil0fil.cc

```diff
@@ -6283,10 +6287,14 @@ fil_load_status Fil_shard::ibd_open_for_recovery(space_id_t space_id,
 #endif /* !UNIV_HOTBACKUP */
 
   const byte *first_page = df.get_first_page();
-  fil_space_crypt_t *crypt_data =
-      first_page
-          ? fil_space_read_crypt_data(page_size_t(df.flags()), first_page)
-          : NULL;
+  fil_space_crypt_t *crypt_data{nullptr};
+  if (df.m_crypt_data) {
+    // crypt data was already read by validate_first_page
+    crypt_data = df.m_crypt_data;
+    df.m_crypt_data = nullptr;
+  } else if (first_page) {
+    crypt_data = fil_space_read_crypt_data(page_size_t(df.flags()), first_page);
```

Comment:
What if crypt_data is only in redo? Where do we read it?

Reply:
log0recv.cc (fil_parse_write_crypt_data_v3)

Reply:
Thanks